csrc/scheduler/registry.cpp

```diff
@@ -748,35 +748,128 @@ bool reductionInterferingView(
   return false;
 }
 
+PrimDataType getTensorIndexType(TensorView* tv, ExpressionEvaluator& ee) {
+  auto non_contig = std::any_of(
+      tv->domain()->contiguity().begin(),
+      tv->domain()->contiguity().end(),
+      [](const auto contig) { return contig.has_value() && !contig.value(); });
+
+  // When a fusion output is non-contiguous, currently there's no
+  // way to obtain its strides. This is an interface problem and
+  // should be fixed.
+  if (tv->isFusionOutput() && non_contig) {
+    return PrimDataType::Int;
+  }
+
+  // Intermediate tensors should be contiguous. There must be
+  // something wrong if not.
+  TORCH_INTERNAL_ASSERT(
+      !non_contig, "Unexpected non-contiguous tensor found: ", tv->toString());
+
+  // Note that at this point tensors are not scheduled yet. Each
+  // tensor may end up being inlined, stored on Shared or Local, but
+  // the index type is currently supposed to be determined before
+  // any of scheduling decisions is made, so we may end up making a
+  // conservative decision.
+  // TODO: Consider index type resolution after segmentation and
+  // scheduling. At that point we have the final scheduled fusions
```

Comment:
Those two functions could make heuristics decisions based the index type. For example, for persistent heuristics, estimating register usage is important, and to improve the accuracy of the estimation, it may make sense to use the information of the index type. For example, in the outer-norm heuristics, we currently use some register factor as a constant overhead, which is then added to the estimate based on the persistent buffer size. I think we currently just have a single constant overhead factor, but maybe this could be improved by considering the index type is 32 or 64 bits.

Another case is more of an artifact of an implementation limitation, and I suspect it wouldn't be common, but it may affect the vectorization factor of a fusion. When we compute the vectorization factor, we look at all the vectorized inputs and outputs, and the tensor with the largest data type dictates the vectorization factor. So for example, if a fusion has half and float inputs, even though the half input could be vectorized by a factor 8, the vectorization factor would be 4 because there's the float input. The index type may be relevant here as it's used for Welford ops, and the count input and output are of type index type, so if a fusion is segmented with a Welford op at a segment boundary, it might affect the vectorization factor.

The second case isn't a fundamental problem but rather it's something we need to fix anyway, but is the concern like the first one important? I don't know. Maybe we could just use int64_t and conservative estimates may be just fine.

Reply:
I see why index type can be relevant to segmentation decision. Thanks for explaining. So we might encounter problems like: if we segment this way, then we will index 64, but if we segment that way, then we will index 32. And in a perfect world, we should take this into consideration and pick the best segmentation. But yeah, considering this in the decision can be complicated, and probably we should make decision assuming 64bit indexing.